language/evm/stdlib/sources/ExternalResult.move

```diff
@@ -0,0 +1,105 @@
+    /// panic_code stores the error code if the error was caused by a panic
+    /// err_data stores the error information as dynamic byte array (vector<u8>)
+    struct ExternalResult<T> has copy, drop, store {
+        value: Option<T>,
+        err_data: Option<vector<u8>>,
+        err_reason: Option<vector<u8>>,
+        panic_code: Option<U256>
+    }
+    spec ExternalResult {
+        /// `Result` cannot contain both a value and an error value.
+        invariant Option::is_some(value) ==> (Option::is_none(err_data) && Option::is_none(err_reason) && Option::is_none(panic_code));
+        invariant Option::is_some(err_data) ==> (Option::is_none(value) && Option::is_none(err_reason) && Option::is_none(panic_code));
+        invariant Option::is_some(err_reason) ==> (Option::is_none(value) && Option::is_none(err_data) && Option::is_none(panic_code));
+        invariant Option::is_some(panic_code) ==> (Option::is_none(value) && Option::is_none(err_data) && Option::is_none(err_reason));
+    }
+
+    /// Return a Result containing `value`.
+    public fun ok<T>(value: T): ExternalResult<T> {
+        ExternalResult<T>{value: Option::some(value), err_data: Option::none<vector<u8>>(), err_reason: Option::none<vector<u8>>(),
+        panic_code: Option::none<U256>()}
+    }
+
+    /// Return a Result containing `err_data`.
+    public fun err_data<T>(error: vector<u8>): ExternalResult<T> {
+        ExternalResult<T>{value: Option::none<T>(), err_data: Option::some(error), err_reason: Option::none<vector<u8>>(),
+        panic_code: Option::none<U256>()}
+    }
+
+    /// Return a Result containing `err_reason`.
+    public fun err_reason<T>(error: vector<u8>): ExternalResult<T> {
+        ExternalResult<T>{value: Option::none<T>(), err_reason: Option::some(error), err_data: Option::none<vector<u8>>(),
+        panic_code: Option::none<U256>()}
+    }
+
+    /// Return a Result containing `panic_code`.
+    public fun panic<T>(panic_code: U256): ExternalResult<T> {
+        ExternalResult<T>{value: Option::none<T>(), err_reason: Option::none<vector<u8>>(), err_data: Option::none<vector<u8>>(),
+        panic_code: Option::some(panic_code)}
+    }
+
+    /// Return true if `result` holds a value.
+    public fun is_ok<T>(result: &ExternalResult<T>): bool {
+        Option::is_some(&result.value)
+    }
+
+    /// Return true if `result` holds an err_data.
+    public fun is_err_data<T>(result: &ExternalResult<T>): bool {
+        Option::is_some(&result.err_data)
+    }
+
+    /// Return true if `result` holds an err_reason.
+    public fun is_err_reason<T>(result: &ExternalResult<T>): bool {
+        Option::is_some(&result.err_reason)
+    }
+
+    /// Return true if `result` holds a panic_code.
+    public fun is_panic<T>(result: &ExternalResult<T>): bool {
+        Option::is_some(&result.panic_code)
+    }
+
+    /// Destroy `result` and extract `value`.
+    public fun unwrap_value<T>(result: ExternalResult<T>): T {
```

Comment:
Another minor comment. "unwrap_value " --> "unwrap"?, just to match to the names of Rust's Result API.